File: apps/web/src/lib/api/vector.ts
/**
 * Vector API: health, index, clear index, search.
 */

import { apiRequest } from "./client";

export interface VectorHealthResponse {
  status: string;
  provider?: string;
  collection_name?: string;
  persist_directory?: string;
  embed_model?: string;
  error?: string;
}

export interface IndexRequestOptions {
  path?: string | null;
  recursive?: boolean;
  reset_collection?: boolean;
  include_chunks?: boolean;
}

export interface IndexDocumentResult {
  source_id: string;
  file_name: string;
  chunk_count: number;
  indexed_chunk_count: number;
  metadata: Record<string, unknown>;
}

export interface IndexResponse {
  status: string;
  collection_name: string;
  total_files: number;
  processed_files: number;
  skipped_files: number;
  indexed_chunks: number;
  documents: IndexDocumentResult[];
  errors: string[];
}

export interface ClearIndexResponse {
  status: string;
  message: string;
  collection_name: string;
}

export async function getVectorHealth(): Promise<VectorHealthResponse> {
  return apiRequest<VectorHealthResponse>("/vector/health", { method: "GET" });
}

export async function postVectorIndex(
  options: IndexRequestOptions = {}
): Promise<IndexResponse> {
  return apiRequest<IndexResponse>("/vector/index", {
    method: "POST",
    body: JSON.stringify({
      path: options.path ?? undefined,
      recursive: options.recursive ?? true,
      reset_collection: options.reset_collection ?? false,
      include_chunks: options.include_chunks ?? false,
    }),
  });
}

export async function deleteVectorIndex(): Promise<ClearIndexResponse> {
  return apiRequest<ClearIndexResponse>("/vector/index", {
    method: "DELETE",
  });
}

File: apps/api/app/schemas/vector.py
"""
Request/response schemas for vector indexing and semantic search.
"""
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class IndexRequest(BaseModel):
    """POST /vector/index body."""

    path: Optional[str] = Field(None, description="Override docs base path")
    recursive: bool = Field(True, description="Scan subdirectories")
    reset_collection: bool = Field(False, description="Clear collection before indexing")
    include_chunks: bool = Field(False, description="Include chunk details in response")


class IndexedChunkResult(BaseModel):
    """One indexed chunk in index response."""

    chunk_id: str
    chunk_index: int
    file_name: str
    source_id: str
    char_count: int
    metadata: Dict[str, Any] = Field(default_factory=dict)


class IndexDocumentResult(BaseModel):
    """One indexed document in index response."""

    source_id: str
    file_name: str
    chunk_count: int
    indexed_chunk_count: int
    metadata: Dict[str, Any] = Field(default_factory=dict)


class IndexResponse(BaseModel):
    """Response for POST /vector/index."""

    status: str
    collection_name: str
    total_files: int = Field(..., ge=0)
    processed_files: int = Field(..., ge=0)
    skipped_files: int = Field(..., ge=0)
    indexed_chunks: int = Field(..., ge=0)
    documents: List[IndexDocumentResult] = Field(default_factory=list)
    errors: List[str] = Field(default_factory=list)


class SearchRequest(BaseModel):
    """POST /vector/search body."""

    query: str = Field(..., min_length=1)
    top_k: Optional[int] = Field(None, ge=1, le=100, description="Override default top_k")
    path_filter: Optional[str] = Field(None, description="Future: filter by path")
    include_text: bool = Field(True, description="Include chunk text in matches")


class SearchMatch(BaseModel):
    """One search result chunk."""

    chunk_id: str
    source_id: str
    file_name: str
    chunk_index: int
    score: float = Field(..., description="Similarity score (higher = more similar)")
    text: str = Field("", description="Chunk text when include_text=true")
    metadata: Dict[str, Any] = Field(default_factory=dict)


class SearchResponse(BaseModel):
    """Response for POST /vector/search."""

    status: str
    query: str
    collection_name: str
    match_count: int = Field(..., ge=0)
    matches: List[SearchMatch] = Field(default_factory=list)
